### PR TITLE
Add support for cross join

### PR DESCRIPTION
**Pull request by joy-yao**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/PreCrossJoinOperator.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+import com.facebook.presto.spi.type.Type;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+@ThreadSafe
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/PreCrossJoinOperator.java]: I don't think this is correct: the class is not thread-safe and `Operator`s do not need to be.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/JoinProbeCompiler.java [hunk 2]
```diff
@@ -129,13 +131,21 @@ public HashJoinOperatorFactoryFactory internalCompileJoinOperatorFactory(List<Ty
...
 
-        Class<? extends OperatorFactory> operatorFactoryClass = IsolatedClass.isolateClass(
-                classLoader,
-                OperatorFactory.class,
-                LookupJoinOperatorFactory.class,
-                LookupJoinOperator.class);
-
-        return new HashJoinOperatorFactoryFactory(joinProbeFactory, operatorFactoryClass);
+        Class<? extends OperatorFactory> operatorFactoryClass =
+                joinType == JoinType.CROSS ?
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/JoinProbeCompiler.java]: For multi-line statements, use an `if` statement rather than ternary.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/JoinProbeCompiler.java [hunk 3]
```diff
@@ -129,13 +131,21 @@ public HashJoinOperatorFactoryFactory internalCompileJoinOperatorFactory(List<Ty
...
+                        IsolatedClass.isolateClass(
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/JoinProbeCompiler.java]: Static import `isolateClass`

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 4]
```diff
@@ -3370,10 +3370,19 @@ public void testCrossJoins()
     }
 
     @Test
-    public void testSimpleCrossJoins()
+    public void testCrossJoinsDifferentDateTypes()
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Keep the original test in addition to this new one

Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Typo in method name: "Date"

File: presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        }
+
+        pageBuilder.declarePosition();
+        writeRow(probeRow, lookupRows.get(lookupPosition), pageBuilder);
+        return !pageBuilder.isFull();
+    }
+
+    private boolean advanceToNextPosition()
+    {
+        if (lookupPosition == lookupSource.getNumberOfRows() - 1 && probePosition == probePage.getPositionCount() - 1) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java]: Add parens

File: presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        // Cannot close lookupSource when the nested loop join is done in parallel
+        lookupSource = null;
+    }
+
+    private void calculateProbeRow()
+    {
+        probeRow = new ArrayList<>();
+        for (int channel = 0; channel < probePage.getChannelCount(); channel++) {
+            Block block = probePage.getBlock(channel);
+            probeRow.add(block.isNull(probePosition) ? null : BlockUtils.convertToType(block.getSlice(probePosition, 0, block.getLength(probePosition)), probeTypes.get(channel)));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java]: This is a bit long. How about
```java
Object row = null;
if (!block.isNull(probePosition)) {
    Slice slice = block.getSlice(probePosition, 0, block.getLength(probePosition));
    row = convertToType(slice, probeTypes.get(channel)));
}

File: presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        }
+    }
+
+    private void writeRow(List<Object> probeRow, List<Object> lookupRow, PageBuilder pageBuilder)
+    {
+        for (int i = 0; i < probeRow.size(); i++) {
+            BlockUtils.appendObject(probeTypes.get(i), pageBuilder.getBlockBuilder(i), probeRow.get(i));
+        }
+
+        for (int i = 0; i < lookupRow.size(); i++) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java]:
```java
for (int i = numProbeColumns; i < (numProbeColumns + lookupRow.size()); i++) {

File: presto-main/src/main/java/com/facebook/presto/block/BlockUtils.java [hunk 8]
```diff
@@ -55,4 +57,32 @@ else if (value instanceof String) {
             throw new IllegalArgumentException("Unsupported type: " + value.getClass().getName());
         }
     }
+
+    public static Object convertToType(Slice slice, Type type)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/block/BlockUtils.java]: There has to be a better way to do this. We shouldn't be adding (more) code to the engine that only works for certain types.

**Comment by electrum** [comment]
This commit message is misleading, since Presto already supports cross join. How about something like "Optimize execution of cross join"

File: presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.facebook.presto.util.MoreFutures.tryGetUnchecked;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+public class NestedLoopJoinOperator
```
Comment (yan-stein) [inline presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java]: @joy-yao It reminds me of choosing appropriate join methods for different join types.  But looks this NLJ operator is introduced for CrossJoin only.  Could you please confirm it?  Thanks.

Comment (joy-yao) [inline presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java]: @yan-stein , yes, currently NestedLoopJoinOperator is only used by CrossJoin.

**Comment by joy-yao** [comment]
Addressed most of David's comments. Waiting for @dain to review.

File: presto-main/src/main/java/com/facebook/presto/block/BlockUtils.java [hunk 10]
```diff
@@ -55,4 +57,32 @@ else if (value instanceof String) {
...
+    {
+        if (slice == null) {
+            return slice;
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/block/BlockUtils.java]: should be `return null`

File: presto-main/src/main/java/com/facebook/presto/operator/JoinOperators.java [hunk 11]
```diff
@@ -15,20 +15,22 @@
 
 import com.facebook.presto.spi.type.Type;
 import com.facebook.presto.sql.gen.JoinProbeCompiler;
+import com.google.common.collect.Lists;
 
 import java.util.List;
 import java.util.Optional;
 
-public class LookupJoinOperators
+public class JoinOperators
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/JoinOperators.java]: Please do the rename in a separate commit before this one.

File: presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+    private final int numProbeColumns;
+
+    private PagesIndexLookupSource lookupSource;
+    private Page probePage;
+
+    private boolean finishing;
+    private int probePosition = -1;
+    private int lookupPosition = -1;
+
+    private List<Object> probeRow = null;
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java]: don't initialize fields to java default values

File: presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+        checkState(lookupSource != null, "Lookup source has not been built yet");
+        checkState(probePage == null, "Current page has not been completely processed yet");
+
+        probePage = page;
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (lookupSource == null) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java]: you don't need this check.  If you have a probe page, you are guaranteed to have a lookup source.

**Comment by dain** [comment]
I think we can simplify this dramatically by replacing the design to cross whole pages instead of row by row.  I'm thinking of something like this:
```
for probePage in probePageStream
   for buildPage in buildPagesList
       outputPages(cross(probePage, buildPage))

crossPage(Page probe, Page build) {
  List<Pages> output;
  for position in probe
     Page probeRow = probe.getSingleValuedPage(position);
     Page probeRun = createRunLengthPage(probeRow, build.positionCount());
     Page outputPage = concatColumns(probeRun, build);
     output(outputPage)
```

Do you follow?

**Comment by dain** [comment]
@joy-yao as part of reviewing this, I did a rebase to current trunk.  Here is the code https://github.com/dain/presto/tree/pr2925

**Comment by dain** [comment]
With the new implementation, we don't need the first commit anymore

**Comment by joy-yao** [comment]
Thanks Dain! I've modified the commits, please review again.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java [hunk 14]
```diff
@@ -305,6 +302,7 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                     newJoinPredicate = joinPredicate; // Use the same as the original
                     break;
                 case FULL:
+                case CROSS:
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: For the CROSS join, i think we should be able push all the *deterministic* predicates down that can be rewritten entirely in terms of each side of the join.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java [hunk 15]
```diff
@@ -617,8 +613,8 @@ private JoinNode tryNormalizeToInnerJoin(JoinNode node, Expression inheritedPred
         {
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: i think this method would better be renamed as tryNormalizeOuterToInnerJoin

Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: i think this method would now better be renamed as tryNormalizeOuterToInnerJoin

File: presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java (hunk 12 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java]: IDEA has an inspection for this.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java [hunk 16]
```diff
@@ -367,7 +362,12 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                     rightSource = new ProjectNode(idAllocator.getNextId(), rightSource, rightProjections.build());
                     criteria = builder.build();
                 }
-                output = new JoinNode(node.getId(), node.getType(), leftSource, rightSource, criteria, node.getLeftHashSymbol(), node.getRightHashSymbol());
+                if (rewriteToCrossJoin) {
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: I don't think you need to special check for cross join above. just do the check here to see if criteria is empty before deciding to do cross join.

**Comment by joy-yao** [comment]
@dain , Eric has reviewed the predicate pushdown. Please review again. Thank you!

**Comment by dain** [comment]
All of my comments were on the first commit.  Can you add the changes as a new commit so it is easier to review, and then we can squash it before merging?

**Comment by joy-yao** [comment]
Thanks @dain, I've added two more commits, one to address the refactoring (rename classes and methods), and the other addresses all other comments. Please review.

**Comment by dain** [comment]
Looks good.  Remember to squash that final commit.
